// eventregistry/dist/queryEvents.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequestEventsRecentActivity = exports.RequestEventsCategoryAggr = exports.RequestEventsEventClusters = exports.RequestEventsDateMentionAggr = exports.RequestEventsSourceAggr = exports.RequestEventsConceptTrends = exports.RequestEventsConceptMatrix = exports.RequestEventsConceptGraph = exports.RequestEventsConceptAggr = exports.RequestEventsLocTimeAggr = exports.RequestEventsLocAggr = exports.RequestEventsKeywordAggr = exports.RequestEventsTimeAggr = exports.RequestEventsUriWgtList = exports.RequestEventsInfo = exports.RequestEvents = exports.QueryEventsIter = exports.QueryEvents = void 0;
var base_1 = require("./base");
var query_1 = require("./query");
var returnInfo_1 = require("./returnInfo");
var logger_1 = require("./logger");
/**
 * @class QueryEvents
 * Class for obtaining available info for one or more events in the Event Registry
 */
var QueryEvents = /** @class */ (function (_super) {
    __extends(QueryEvents, _super);
    function QueryEvents(args) {
        if (args === void 0) { args = {}; }
        var _this = _super.call(this) || this;
        _this.params = {};
        var keywords = args.keywords, conceptUri = args.conceptUri, categoryUri = args.categoryUri, sourceUri = args.sourceUri, sourceLocationUri = args.sourceLocationUri, sourceGroupUri = args.sourceGroupUri, authorUri = args.authorUri, locationUri = args.locationUri, lang = args.lang, dateStart = args.dateStart, dateEnd = args.dateEnd, reportingDateStart = args.reportingDateStart, reportingDateEnd = args.reportingDateEnd, _a = args.minSentiment, minSentiment = _a === void 0 ? -1 : _a, _b = args.maxSentiment, maxSentiment = _b === void 0 ? 1 : _b, minArticlesInEvent = args.minArticlesInEvent, maxArticlesInEvent = args.maxArticlesInEvent, dateMentionStart = args.dateMentionStart, dateMentionEnd = args.dateMentionEnd, ignoreKeywords = args.ignoreKeywords, ignoreConceptUri = args.ignoreConceptUri, ignoreCategoryUri = args.ignoreCategoryUri, ignoreSourceUri = args.ignoreSourceUri, ignoreSourceLocationUri = args.ignoreSourceLocationUri, ignoreSourceGroupUri = args.ignoreSourceGroupUri, ignoreAuthorUri = args.ignoreAuthorUri, ignoreLocationUri = args.ignoreLocationUri, ignoreLang = args.ignoreLang, _c = args.keywordsLoc, keywordsLoc = _c === void 0 ? "body" : _c, _d = args.ignoreKeywordsLoc, ignoreKeywordsLoc = _d === void 0 ? "body" : _d, _e = args.keywordSearchMode, keywordSearchMode = _e === void 0 ? "phrase" : _e, _f = args.ignoreKeywordSearchMode, ignoreKeywordSearchMode = _f === void 0 ? "phrase" : _f, _g = args.categoryIncludeSub, categoryIncludeSub = _g === void 0 ? true : _g, _h = args.ignoreCategoryIncludeSub, ignoreCategoryIncludeSub = _h === void 0 ? true : _h, _j = args.requestedResult, requestedResult = _j === void 0 ? new RequestEventsInfo() : _j;
        _this.setVal("action", "getEvents");
        _this.setQueryArrVal(keywords, "keyword", "keywordOper", "and");
        _this.setQueryArrVal(conceptUri, "conceptUri", "conceptOper", "and");
        _this.setQueryArrVal(categoryUri, "categoryUri", "categoryOper", "or");
        _this.setQueryArrVal(sourceUri, "sourceUri", undefined, "or");
        _this.setQueryArrVal(sourceLocationUri, "sourceLocationUri", undefined, "or");
        _this.setQueryArrVal(sourceGroupUri, "sourceGroupUri", undefined, "or");
        _this.setQueryArrVal(authorUri, "authorUri", "authorOper", "or");
        _this.setQueryArrVal(locationUri, "locationUri", undefined, "or");
        _this.setQueryArrVal(lang, "lang", undefined, "or");
        if (dateStart !== undefined) {
            _this.setDateVal("dateStart", dateStart);
        }
        if (dateEnd !== undefined) {
            _this.setDateVal("dateEnd", dateEnd);
        }
        if (reportingDateStart !== undefined) {
            _this.setDateVal("reportingDateStart", reportingDateStart);
        }
        if (reportingDateEnd !== undefined) {
            _this.setDateVal("reportingDateEnd", reportingDateEnd);
        }
        if (minSentiment > -1 && minSentiment <= 1) {
            _this.setVal("minSentiment", minSentiment);
        }
        if (maxSentiment >= -1 && maxSentiment < 1) {
            _this.setVal("maxSentiment", maxSentiment);
        }
        _this.setValIfNotDefault("minArticlesInEvent", minArticlesInEvent, undefined);
        _this.setValIfNotDefault("maxArticlesInEvent", maxArticlesInEvent, undefined);
        if (dateMentionStart !== undefined) {
            _this.setDateVal("dateMentionStart", dateMentionStart);
        }
        if (dateMentionEnd !== undefined) {
            _this.setDateVal("dateMentionEnd", dateMentionEnd);
        }
        // for the negative conditions, only the OR is a valid operator type
        _this.setQueryArrVal(ignoreKeywords, "ignoreKeywords", undefined, "or");
        _this.setQueryArrVal(ignoreConceptUri, "ignoreConceptUri", undefined, "or");
        _this.setQueryArrVal(ignoreCategoryUri, "ignoreCategoryUri", undefined, "or");
        _this.setQueryArrVal(ignoreSourceUri, "ignoreSourceUri", undefined, "or");
        _this.setQueryArrVal(ignoreSourceLocationUri, "ignoreSourceLocationUri", undefined, "or");
        _this.setQueryArrVal(ignoreSourceGroupUri, "ignoreSourceGroupUri", undefined, "or");
        _this.setQueryArrVal(ignoreAuthorUri, "ignoreAuthorUri", undefined, "or");
        _this.setQueryArrVal(ignoreLocationUri, "ignoreLocationUri", undefined, "or");
        _this.setQueryArrVal(ignoreLang, "ignoreLang", undefined, "or");
        _this.setValIfNotDefault("keywordLoc", keywordsLoc, "body");
        _this.setValIfNotDefault("ignoreKeywordLoc", ignoreKeywordsLoc, "body");
        _this.setValIfNotDefault("keywordSearchMode", keywordSearchMode, "phrase");
        _this.setValIfNotDefault("ignoreKeywordSearchMode", ignoreKeywordSearchMode, "phrase");
        _this.setValIfNotDefault("categoryIncludeSub", categoryIncludeSub, true);
        _this.setValIfNotDefault("ignoreCategoryIncludeSub", ignoreCategoryIncludeSub, true);
        _this.setRequestedResult(requestedResult);
        return _this;
    }
    Object.defineProperty(QueryEvents.prototype, "path", {
        get: function () {
            return "/api/v1/event";
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Set the single result type that you would like to be returned. Any previously set result types will be overwritten.
     * Result types can be the classes that extend RequestEvent base class (see classes below).
     */
    QueryEvents.prototype.setRequestedResult = function (requestEvents) {
        if (!(requestEvents instanceof RequestEvents)) {
            throw new Error("QueryEvents class can only accept result requests that are of type RequestEvents");
        }
        this.resultTypeList = [
            requestEvents,
        ];
    };
    /**
     * Set a custom list of event uris. The results will be then computed on this list - no query will be done (all conditions will be ignored).
     */
    QueryEvents.initWithEventUriList = function (uriList) {
        var query = new QueryEvents();
        if (!Array.isArray(uriList)) {
            throw new Error("uriList has to be a list of strings that represent event uris");
        }
        query.params = {
            action: "getEvents",
            eventUriList: uriList.join(","),
        };
        return query;
    };
    QueryEvents.initWithEventUriWgtList = function (uriWgtList) {
        var query = new QueryEvents();
        if (!Array.isArray(uriWgtList)) {
            throw new Error("uriWgtList has to be a list of strings that represent event uris");
        }
        query.params = {
            action: "getEvents",
            eventUriWgtList: uriWgtList.join(","),
        };
        return query;
    };
    QueryEvents.initWithComplexQuery = function (complexQuery) {
        var query = new QueryEvents();
        if (complexQuery instanceof query_1.ComplexEventQuery) {
            query.setVal("query", JSON.stringify(complexQuery.getQuery()));
        }
        else if (typeof complexQuery === 'string') {
            try {
                JSON.parse(complexQuery);
            }
            catch (_a) {
                logger_1.Logger.error("Failed to parse the provided string content as a JSON object. Please check the content provided as a parameter to the initWithComplexQuery() method");
            }
            query.setVal("query", complexQuery);
        }
        else if (typeof complexQuery === 'object' && complexQuery !== null) {
            query.setVal("query", JSON.stringify(complexQuery));
        }
        else {
            throw new Error("The instance of query parameter was not a ComplexEventQuery, a string or an object");
        }
        return query;
    };
    return QueryEvents;
}(base_1.Query));
exports.QueryEvents = QueryEvents;
/**
 * @class QueryEventsIter
 * Class for iterating through all the events via callbacks
 */
var QueryEventsIter = /** @class */ (function (_super) {
    __extends(QueryEventsIter, _super);
    function QueryEventsIter(er, args) {
        if (args === void 0) { args = {}; }
        var _this = _super.call(this, args) || this;
        _this.page = 0;
        _this.pages = 1;
        _this.items = [];
        _this.returnedSoFar = 0;
        _this.index = 0;
        _this.callback = function () { return undefined; };
        _this.doneCallback = function () { return undefined; };
        var _a = args.sortBy, sortBy = _a === void 0 ? "rel" : _a, _b = args.sortByAsc, sortByAsc = _b === void 0 ? false : _b, _c = args.returnInfo, returnInfo = _c === void 0 ? undefined : _c, _d = args.maxItems, maxItems = _d === void 0 ? -1 : _d;
        _this.er = er;
        _this.sortBy = sortBy;
        _this.sortByAsc = sortByAsc;
        _this.returnInfo = returnInfo;
        _this.maxItems = maxItems;
        return _this;
    }
    /**
     * Async Iterator function that returns the next item in the list of events
     */
    QueryEventsIter.prototype[Symbol.asyncIterator] = function () {
        var _this = this;
        return {
            next: function () { return __awaiter(_this, void 0, void 0, function () {
                var item;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (!(this.index >= this.items.length)) return [3 /*break*/, 2];
                            return [4 /*yield*/, this.getNextBatch()];
                        case 1:
                            _a.sent();
                            _a.label = 2;
                        case 2:
                            item = this.items[this.index];
                            this.index++;
                            return [2 /*return*/, { value: item, done: !item }];
                    }
                });
            }); },
        };
    };
    QueryEventsIter.prototype.count = function () {
        return __awaiter(this, void 0, void 0, function () {
            var response;
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        this.setRequestedResult(new RequestEventsInfo());
                        return [4 /*yield*/, this.er.execQuery(this)];
                    case 1:
                        response = _b.sent();
                        if (response === null || response === void 0 ? void 0 : response.error) {
                            this.er.logger.error(response.error);
                        }
                        return [2 /*return*/, ((_a = response === null || response === void 0 ? void 0 : response.events) === null || _a === void 0 ? void 0 : _a.totalResults) || 0];
                }
            });
        });
    };
    /**
     * Execute query and fetch batches of events of the specified size (eventBatchSize)
     * @param callback callback function that'll be called every time we get a new batch of events from the backend
     * @param doneCallback callback function that'll be called when everything is complete
     */
    QueryEventsIter.prototype.execQuery = function (callback, doneCallback) {
        if (callback) {
            this.callback = callback;
        }
        if (doneCallback) {
            this.doneCallback = doneCallback;
        }
        this.iterate();
    };
    QueryEventsIter.initWithComplexQuery = function (er, complexQuery, args) {
        if (args === void 0) { args = {}; }
        var query = new QueryEventsIter(er, args);
        if (complexQuery instanceof query_1.ComplexEventQuery) {
            query.setVal("query", JSON.stringify(complexQuery.getQuery()));
        }
        else if (typeof complexQuery === "string") {
            query.setVal("query", complexQuery);
        }
        else if (typeof complexQuery === "object" && complexQuery !== null) {
            query.setVal("query", JSON.stringify(complexQuery));
        }
        else {
            throw new Error("The instance of query parameter was not a ComplexEventQuery, a string or an object");
        }
        return query;
    };
    QueryEventsIter.prototype.iterate = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.current) return [3 /*break*/, 1];
                        this.callback(this.current);
                        this.index += 1;
                        return [3 /*break*/, 3];
                    case 1: return [4 /*yield*/, this.getNextBatch()];
                    case 2:
                        if (!(_a.sent())) {
                            this.doneCallback(this.errorMessage);
                            return [2 /*return*/];
                        }
                        _a.label = 3;
                    case 3: return [2 /*return*/, this.iterate()];
                }
            });
        });
    };
    /**
     * Extract the results according to maxItems
     * @param response response from the backend
     */
    QueryEventsIter.prototype.extractResults = function (response) {
        var _a;
        var results = ((_a = response === null || response === void 0 ? void 0 : response.events) === null || _a === void 0 ? void 0 : _a.results) || [];
        var extractedSize = this.maxItems !== -1 ? this.maxItems - this.returnedSoFar : results.length;
        return results.slice(0, extractedSize).filter(Boolean);
    };
    Object.defineProperty(QueryEventsIter.prototype, "current", {
        get: function () {
            return this.items[this.index] || undefined;
        },
        enumerable: false,
        configurable: true
    });
    QueryEventsIter.prototype.getNextBatch = function () {
        return __awaiter(this, void 0, void 0, function () {
            var requestEventsInfo, response, error, results, error_1;
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 2, , 3]);
                        this.page += 1;
                        if (this.page > this.pages || (this.maxItems !== -1 && this.returnedSoFar >= this.maxItems)) {
                            return [2 /*return*/, false];
                        }
                        requestEventsInfo = new RequestEventsInfo({
                            page: this.page,
                            count: 50,
                            sortBy: this.sortBy,
                            sortByAsc: this.sortByAsc,
                            returnInfo: this.returnInfo,
                        });
                        this.setRequestedResult(requestEventsInfo);
                        if (this.er.verboseOutput) {
                            this.er.logger.info("Downloading event page ".concat(this.page, "..."));
                        }
                        return [4 /*yield*/, this.er.execQuery(this, this.er.allowUseOfArchive)];
                    case 1:
                        response = _b.sent();
                        error = response.error || "";
                        if (error) {
                            this.errorMessage = "Error while obtaining a list of events:  ".concat(response.error);
                        }
                        else {
                            this.pages = ((_a = response.events) === null || _a === void 0 ? void 0 : _a.pages) || 0;
                        }
                        results = this.extractResults(response);
                        this.returnedSoFar += results.length;
                        this.items = __spreadArray(__spreadArray([], __read(this.items), false), __read(results), false);
                        return [2 /*return*/, true];
                    case 2:
                        error_1 = _b.sent();
                        this.er.logger.error(error_1);
                        return [2 /*return*/, false];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    return QueryEventsIter;
}(QueryEvents));
exports.QueryEventsIter = QueryEventsIter;
var RequestEvents = /** @class */ (function () {
    function RequestEvents() {
    }
    return RequestEvents;
}());
exports.RequestEvents = RequestEvents;
/**
 * @class RequestEventsInfo
 * Return event details for resulting events.
 */
var RequestEventsInfo = /** @class */ (function (_super) {
    __extends(RequestEventsInfo, _super);
    function RequestEventsInfo(args) {
        if (args === void 0) { args = {}; }
        var _this = _super.call(this) || this;
        _this.resultType = "events";
        var _a = args.page, page = _a === void 0 ? 1 : _a, _b = args.count, count = _b === void 0 ? 50 : _b, _c = args.sortBy, sortBy = _c === void 0 ? "rel" : _c, _d = args.sortByAsc, sortByAsc = _d === void 0 ? false : _d, _e = args.returnInfo, returnInfo = _e === void 0 ? undefined : _e;
        if (page < 1) {
            throw new RangeError("Page has to be >= 1");
        }
        if (count > 50) {
            throw new RangeError("At most 50 events can be returned per call");
        }
        _this.params = {};
        _this.params["eventsPage"] = page;
        _this.params["eventsCount"] = count;
        _this.params["eventsSortBy"] = sortBy;
        _this.params["eventsSortByAsc"] = sortByAsc;
        if (!!returnInfo) {
            _this.params = __assign(__assign({}, _this.params), returnInfo.getParams("events"));
        }
        return _this;
    }
    return RequestEventsInfo;
}(RequestEvents));
exports.RequestEventsInfo = RequestEventsInfo;
/**
 * @class RequestEventsUriList
 * Return a simple list of event uris together with the scores for resulting events.
 */
var RequestEventsUriWgtList = /** @class */ (function (_super) {
    __extends(RequestEventsUriWgtList, _super);
    function RequestEventsUriWgtList(args) {
        if (args === void 0) { args = {}; }
        var _this = _super.call(this) || this;
        _this.resultType = "uriWgtList";
        var _a = args.page, page = _a === void 0 ? 1 : _a, _b = args.count, count = _b === void 0 ? 50000 : _b, _c = args.sortBy, sortBy = _c === void 0 ? "rel" : _c, _d = args.sortByAsc, sortByAsc = _d === void 0 ? false : _d, unsupported = __rest(args, ["page", "count", "sortBy", "sortByAsc"]);
        if (Object.keys(unsupported).length !== 0) {
            logger_1.Logger.warn("RequestEventsUriWgtList: Unsupported parameters detected: ".concat(JSON.stringify(unsupported), ". Please check the documentation."));
        }
        if (page < 1) {
            throw new RangeError("Page has to be >= 1");
        }
        if (count > 100000) {
            throw new RangeError("At most 100000 results can be returned per call");
        }
        _this.params = {};
        _this.params["uriWgtListPage"] = page;
        _this.params["uriWgtListCount"] = count;
        _this.params["uriWgtListSortBy"] = sortBy;
        _this.params["uriWgtListSortByAsc"] = sortByAsc;
        return _this;
    }
    Object.defineProperty(RequestEventsUriWgtList.prototype, "page", {
        set: function (page) {
            if (page < 1) {
                throw new RangeError("Page has to be >= 1");
            }
            this.params["uriWgtListPage"] = page;
        },
        enumerable: false,
        configurable: true
    });
    return RequestEventsUriWgtList;
}(RequestEvents));
exports.RequestEventsUriWgtList = RequestEventsUriWgtList;
/**
 * @class RequestEventsTimeAggr
 * Return time distribution of resulting events.
 */
var RequestEventsTimeAggr = /** @class */ (function (_super) {
    __extends(RequestEventsTimeAggr, _super);
    function RequestEventsTimeAggr() {
        var _this = _super.apply(this, __spreadArray([], __read(arguments), false)) || this;
        _this.resultType = "timeAggr";
        return _this;
    }
    return RequestEventsTimeAggr;
}(RequestEvents));
exports.RequestEventsTimeAggr = RequestEventsTimeAggr;
/**
 * Return keyword aggregate (tag cloud) on words in articles in resulting events.
 * @class RequestEventsTimeAggr
 */
var RequestEventsKeywordAggr = /** @class */ (function (_super) {
    __extends(RequestEventsKeywordAggr, _super);
    /**
     * @param lang: in which language to produce the list of top keywords. If undefined, then compute on all articles
     */
    function RequestEventsKeywordAggr(lang) {
        var _this = _super.call(this) || this;
        _this.resultType = "keywordAggr";
        _this.params = {};
        if (lang !== undefined) {
            _this.params["keywordAggrLang"] = lang;
        }
        return _this;
    }
    return RequestEventsKeywordAggr;
}(RequestEvents));
exports.RequestEventsKeywordAggr = RequestEventsKeywordAggr;
/**
 * @class RequestEventsLocAggr
 * Return aggregate of locations of resulting events.
 */
var RequestEventsLocAggr = /** @class */ (function (_super) {
    __extends(RequestEventsLocAggr, _super);
    function RequestEventsLocAggr(args) {
        if (args === void 0) { args = {}; }
        var _this = _super.call(this) || this;
        _this.resultType = "locAggr";
        var _a = args.eventsSampleSize, eventsSampleSize = _a === void 0 ? 100000 : _a, _b = args.returnInfo, returnInfo = _b === void 0 ? new returnInfo_1.ReturnInfo() : _b, unsupported = __rest(args, ["eventsSampleSize", "returnInfo"]);
        if (Object.keys(unsupported).length !== 0) {
            logger_1.Logger.warn("RequestEventsLocAggr: Unsupported parameters detected: ".concat(JSON.stringify(unsupported), ". Please check the documentation."));
        }
        if (eventsSampleSize > 100000) {
            throw new RangeError("At most 100000 results can be used for computing");
        }
        _this.params = {};
        _this.params["locAggrSampleSize"] = eventsSampleSize;
        _this.params = __assign(__assign({}, _this.params), returnInfo.getParams("locAggr"));
        return _this;
    }
    return RequestEventsLocAggr;
}(RequestEvents));
exports.RequestEventsLocAggr = RequestEventsLocAggr;
/**
 * @class RequestEventsLocTimeAggr
 * Return aggregate of locations and times of resulting events.
 */
var RequestEventsLocTimeAggr = /** @class */ (function (_super) {
    __extends(RequestEventsLocTimeAggr, _super);
    function RequestEventsLocTimeAggr(args) {
        if (args === void 0) { args = {}; }
        var _this = _super.call(this) || this;
        _this.resultType = "locTimeAggr";
        var _a = args.eventsSampleSize, eventsSampleSize = _a === void 0 ? 100000 : _a, _b = args.returnInfo, returnInfo = _b === void 0 ? new returnInfo_1.ReturnInfo() : _b, unsupported = __rest(args, ["eventsSampleSize", "returnInfo"]);
        if (Object.keys(unsupported).length !== 0) {
            logger_1.Logger.warn("RequestEventsLocTimeAggr: Unsupported parameters detected: ".concat(JSON.stringify(unsupported), ". Please check the documentation."));
        }
        if (eventsSampleSize > 100000) {
            throw new RangeError("At most 100000 results can be used for computing");
        }
        _this.params = {};
        _this.params["locTimeAggrSampleSize"] = eventsSampleSize;
        _this.params = __assign(__assign({}, _this.params), returnInfo.getParams("locTimeAggr"));
        return _this;
    }
    return RequestEventsLocTimeAggr;
}(RequestEvents));
exports.RequestEventsLocTimeAggr = RequestEventsLocTimeAggr;
/**
 * @class RequestEventsConceptAggr
 * Compute which concept are the most frequently occurring in the list of resulting events.
 */
var RequestEventsConceptAggr = /** @class */ (function (_super) {
    __extends(RequestEventsConceptAggr, _super);
    function RequestEventsConceptAggr(args) {
        if (args === void 0) { args = {}; }
        var _this = _super.call(this) || this;
        _this.resultType = "conceptAggr";
        var _a = args.conceptCount, conceptCount = _a === void 0 ? 20 : _a, _b = args.eventsSampleSize, eventsSampleSize = _b === void 0 ? 100000 : _b, _c = args.returnInfo, returnInfo = _c === void 0 ? new returnInfo_1.ReturnInfo() : _c;
        if (conceptCount > 200) {
            throw new RangeError("At most 200 top concepts can be returned");
        }
        if (eventsSampleSize > 300000) {
            throw new RangeError("At most 300000 results can be used for computing");
        }
        _this.params = {};
        _this.params["conceptAggrConceptCount"] = conceptCount;
        _this.params["conceptAggrSampleSize"] = eventsSampleSize;
        _this.params = __assign(__assign({}, _this.params), returnInfo.getParams("conceptAggr"));
        return _this;
    }
    return RequestEventsConceptAggr;
}(RequestEvents));
exports.RequestEventsConceptAggr = RequestEventsConceptAggr;
/**
 * @class RequestEventsConceptGraph
 * Compute which concept pairs frequently co-occur together in the resulting events.
 */
var RequestEventsConceptGraph = /** @class */ (function (_super) {
    __extends(RequestEventsConceptGraph, _super);
    function RequestEventsConceptGraph(args) {
        if (args === void 0) { args = {}; }
        var _this = _super.call(this) || this;
        _this.resultType = "conceptGraph";
        var _a = args.conceptCount, conceptCount = _a === void 0 ? 25 : _a, _b = args.linkCount, linkCount = _b === void 0 ? 150 : _b, _c = args.eventsSampleSize, eventsSampleSize = _c === void 0 ? 50000 : _c, _d = args.returnInfo, returnInfo = _d === void 0 ? new returnInfo_1.ReturnInfo() : _d;
        if (conceptCount > 1000) {
            throw new RangeError("At most 1000 top concepts can be returned");
        }
        if (linkCount > 2000) {
            throw new RangeError("At most 2000 links between concepts can be returned");
        }
        if (eventsSampleSize > 300000) {
            throw new RangeError("At most 300000 results can be used for computing");
        }
        _this.params = {};
        _this.params["conceptGraphConceptCount"] = conceptCount;
        _this.params["conceptGraphLinkCount"] = linkCount;
        _this.params["conceptGraphSampleSize"] = eventsSampleSize;
        _this.params = __assign(__assign({}, _this.params), returnInfo.getParams("conceptGraph"));
        return _this;
    }
    return RequestEventsConceptGraph;
}(RequestEvents));
exports.RequestEventsConceptGraph = RequestEventsConceptGraph;
/**
 * @class RequestEventsConceptMatrix
 * Get a matrix of concepts and their dependencies.
 * For individual concept pairs return how frequently
 * they co-occur in the resulting events and how "surprising" this is,
 * based on the frequency of individual concepts.
 */
var RequestEventsConceptMatrix = /** @class */ (function (_super) {
    __extends(RequestEventsConceptMatrix, _super);
    function RequestEventsConceptMatrix(args) {
        if (args === void 0) { args = {}; }
        var _this = _super.call(this) || this;
        _this.resultType = "conceptMatrix";
        var _a = args.conceptCount, conceptCount = _a === void 0 ? 25 : _a, _b = args.measure, measure = _b === void 0 ? "pmi" : _b, _c = args.eventsSampleSize, eventsSampleSize = _c === void 0 ? 100000 : _c, _d = args.returnInfo, returnInfo = _d === void 0 ? new returnInfo_1.ReturnInfo() : _d;
        if (conceptCount > 200) {
            throw new RangeError("At most 200 top concepts can be returned");
        }
        if (eventsSampleSize > 300000) {
            throw new RangeError("At most 300000 results can be used for computing");
        }
        _this.params = {};
        _this.params["conceptMatrixConceptCount"] = conceptCount;
        _this.params["conceptMatrixMeasure"] = measure;
        _this.params["conceptMatrixSampleSize"] = eventsSampleSize;
        _this.params = __assign(__assign({}, _this.params), returnInfo.getParams("conceptMatrix"));
        return _this;
    }
    return RequestEventsConceptMatrix;
}(RequestEvents));
exports.RequestEventsConceptMatrix = RequestEventsConceptMatrix;
/**
 * @class RequestEventsConceptTrends
 * Return a list of top trending concepts and their daily trending info over time
 */
var RequestEventsConceptTrends = /** @class */ (function (_super) {
    __extends(RequestEventsConceptTrends, _super);
    function RequestEventsConceptTrends(args) {
        if (args === void 0) { args = {}; }
        var _this = _super.call(this) || this;
        _this.resultType = "conceptTrends";
        var _a = args.conceptCount, conceptCount = _a === void 0 ? 10 : _a, _b = args.conceptUris, conceptUris = _b === void 0 ? [] : _b, _c = args.returnInfo, returnInfo = _c === void 0 ? new returnInfo_1.ReturnInfo() : _c;
        if (conceptCount > 50) {
            throw new RangeError("At most 50 top concepts can be returned");
        }
        _this.params = {};
        _this.params["conceptTrendsConceptCount"] = conceptCount;
        if (conceptUris.length > 0) {
            _this.params["conceptTrendsConceptUri"] = conceptUris;
        }
        _this.params = __assign(__assign({}, _this.params), returnInfo.getParams("conceptTrends"));
        return _this;
    }
    return RequestEventsConceptTrends;
}(RequestEvents));
exports.RequestEventsConceptTrends = RequestEventsConceptTrends;
/**
 * @class RequestEventsSourceAggr
 * Return top news sources that report about the events that match the search conditions
 */
var RequestEventsSourceAggr = /** @class */ (function (_super) {
    __extends(RequestEventsSourceAggr, _super);
    function RequestEventsSourceAggr(args) {
        if (args === void 0) { args = {}; }
        var _this = _super.call(this) || this;
        _this.resultType = "sourceAggr";
        var _a = args.sourceCount, sourceCount = _a === void 0 ? 30 : _a, _b = args.eventsSampleSize, eventsSampleSize = _b === void 0 ? 50000 : _b, _c = args.returnInfo, returnInfo = _c === void 0 ? new returnInfo_1.ReturnInfo() : _c;
        if (sourceCount > 200) {
            throw new RangeError("At most 200 top sources can be returned");
        }
        if (eventsSampleSize > 100000) {
            throw new RangeError("At most 100000 results can be used for computing");
        }
        _this.params = {};
        _this.params["sourceAggrSourceCount"] = sourceCount;
        _this.params["sourceAggrSampleSize"] = eventsSampleSize;
        _this.params = __assign(__assign({}, _this.params), returnInfo.getParams("sourceAggr"));
        return _this;
    }
    return RequestEventsSourceAggr;
}(RequestEvents));
exports.RequestEventsSourceAggr = RequestEventsSourceAggr;
/**
 * @class RequestEventsDateMentionAggr
 * Return events and the dates that are mentioned in articles about these events
 */
var RequestEventsDateMentionAggr = /** @class */ (function (_super) {
    __extends(RequestEventsDateMentionAggr, _super);
    function RequestEventsDateMentionAggr(args) {
        if (args === void 0) { args = {}; }
        var _this = _super.call(this) || this;
        _this.resultType = "dateMentionAggr";
        var _a = args.minDaysApart, minDaysApart = _a === void 0 ? 0 : _a, _b = args.minDateMentionCount, minDateMentionCount = _b === void 0 ? 5 : _b, _c = args.eventsSampleSize, eventsSampleSize = _c === void 0 ? 100000 : _c, _d = args.returnInfo, returnInfo = _d === void 0 ? new returnInfo_1.ReturnInfo() : _d;
        if (eventsSampleSize > 300000) {
            throw new RangeError("At most 300000 results can be used for computing");
        }
        _this.params = {};
        _this.params["dateMentionAggrMinDaysApart"] = minDaysApart;
        _this.params["dateMentionAggrMinDateMentionCount"] = minDateMentionCount;
        _this.params["dateMentionAggrSampleSize"] = eventsSampleSize;
        _this.params = __assign(__assign({}, _this.params), returnInfo.getParams("dateMentionAggr"));
        return _this;
    }
    return RequestEventsDateMentionAggr;
}(RequestEvents));
exports.RequestEventsDateMentionAggr = RequestEventsDateMentionAggr;
/**
 * @class RequestEventsEventClusters
 * Return hierarchical clustering of events into smaller clusters. 2-means clustering is applied on each node in the tree
 */
var RequestEventsEventClusters = /** @class */ (function (_super) {
    __extends(RequestEventsEventClusters, _super);
    function RequestEventsEventClusters(args) {
        if (args === void 0) { args = {}; }
        var _this = _super.call(this) || this;
        _this.resultType = "eventClusters";
        var _a = args.keywordCount, keywordCount = _a === void 0 ? 30 : _a, _b = args.maxEventsToCluster, maxEventsToCluster = _b === void 0 ? 10000 : _b, _c = args.returnInfo, returnInfo = _c === void 0 ? new returnInfo_1.ReturnInfo() : _c;
        if (keywordCount > 100) {
            throw new RangeError("At most 100 keywords can be reported in each of the clusters");
        }
        if (maxEventsToCluster > 10000) {
            throw new RangeError("At most 10000 events can be clustered together");
        }
        _this.params = {};
        _this.params["eventClustersKeywordCount"] = keywordCount;
        _this.params["eventClustersMaxEventsToCluster"] = maxEventsToCluster;
        _this.params = __assign(__assign({}, _this.params), returnInfo.getParams("eventClusters"));
        return _this;
    }
    return RequestEventsEventClusters;
}(RequestEvents));
exports.RequestEventsEventClusters = RequestEventsEventClusters;
/**
 * @class RequestEventsCategoryAggr
 * Return distribution of events into dmoz categories.
 */
var RequestEventsCategoryAggr = /** @class */ (function (_super) {
    __extends(RequestEventsCategoryAggr, _super);
    function RequestEventsCategoryAggr(returnInfo) {
        if (returnInfo === void 0) { returnInfo = new returnInfo_1.ReturnInfo(); }
        var _this = _super.call(this) || this;
        _this.resultType = "categoryAggr";
        _this.params = returnInfo.getParams("categoryAggr");
        return _this;
    }
    return RequestEventsCategoryAggr;
}(RequestEvents));
exports.RequestEventsCategoryAggr = RequestEventsCategoryAggr;
/**
 * @class RequestEventsRecentActivity
 * Return a list of recently changed events that match search conditions.
 */
var RequestEventsRecentActivity = /** @class */ (function (_super) {
    __extends(RequestEventsRecentActivity, _super);
    function RequestEventsRecentActivity(args) {
        if (args === void 0) { args = {}; }
        var _this = _super.call(this) || this;
        _this.resultType = "recentActivityEvents";
        var _a = args.maxEventCount, maxEventCount = _a === void 0 ? 50 : _a, updatesAfterTm = args.updatesAfterTm, updatesAfterMinsAgo = args.updatesAfterMinsAgo, _b = args.mandatoryLocation, mandatoryLocation = _b === void 0 ? true : _b, _c = args.minAvgCosSim, minAvgCosSim = _c === void 0 ? 0 : _c, _d = args.returnInfo, returnInfo = _d === void 0 ? undefined : _d;
        if (maxEventCount > 2000) {
            throw new RangeError("At most 2000 events can be returned");
        }
        if (updatesAfterTm !== undefined && updatesAfterMinsAgo !== undefined) {
            throw new Error("You should specify either updatesAfterTm or updatesAfterMinsAgo parameter, but not both");
        }
        _this.params = {};
        _this.params["recentActivityEventsMaxEventCount"] = maxEventCount;
        _this.params["recentActivityEventsMandatoryLocation"] = mandatoryLocation;
        if (updatesAfterTm !== undefined) {
            _this.params["recentActivityEventsUpdatesAfterTm"] = base_1.QueryParamsBase.encodeDateTime(updatesAfterTm);
        }
        if (updatesAfterMinsAgo !== undefined) {
            _this.params["recentActivityEventsUpdatesAfterMinsAgo"] = updatesAfterMinsAgo;
        }
        _this.params["recentActivityEventsMinAvgCosSim"] = minAvgCosSim;
        if (!!returnInfo) {
            _this.params = __assign(__assign({}, _this.params), returnInfo.getParams("recentActivityEvents"));
        }
        return _this;
    }
    return RequestEventsRecentActivity;
}(RequestEvents));
exports.RequestEventsRecentActivity = RequestEventsRecentActivity;
