// eventregistry/dist/returnInfo.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConceptFolderInfoFlags = exports.ConceptClassInfoFlags = exports.LocationInfoFlags = exports.ConceptInfoFlags = exports.CategoryInfoFlags = exports.SourceInfoFlags = exports.EventInfoFlags = exports.StoryInfoFlags = exports.ArticleInfoFlags = exports.ReturnInfo = exports.ReturnInfoFlagsBase = void 0;
var fs = require("fs");
var logger_1 = require("./logger");
var ReturnInfoFlagsBase = /** @class */ (function () {
    function ReturnInfoFlagsBase() {
        this.data = {};
    }
    ReturnInfoFlagsBase.prototype.setFlag = function (key, value, defaultValue) {
        this.setProperty("Include" + this.type + key.charAt(0).toUpperCase() + key.slice(1), value, defaultValue);
    };
    ReturnInfoFlagsBase.prototype.setValue = function (key, value, defaultValue, skipKeyMod) {
        if (skipKeyMod === void 0) { skipKeyMod = false; }
        var constructedKey = skipKeyMod ? key.charAt(0).toUpperCase() + key.slice(1) : this.type + key.charAt(0).toUpperCase() + key.slice(1);
        this.setProperty(constructedKey, value, defaultValue);
    };
    ReturnInfoFlagsBase.prototype.getProperties = function (prefix) {
        if (prefix === void 0) { prefix = ""; }
        var result = {};
        for (var key in this.data) {
            if (Object.prototype.hasOwnProperty.call(this.data, key)) {
                var value = this.data[key];
                var lowercaseKey = key.toLowerCase();
                var lowercasePrefix = prefix.toLowerCase();
                if (lowercaseKey.startsWith(lowercasePrefix)) {
                    result[this.camelCase(key)] = value;
                }
                else {
                    result[this.camelCase(prefix + key)] = value;
                }
            }
        }
        return result;
    };
    ReturnInfoFlagsBase.prototype.camelCase = function (str) {
        return str.replace(/(?:^\w|[A-Z]|\b\w)/g, function (word, index) {
            return index === 0 ? word.toLowerCase() : word.toUpperCase();
        }).replace(/\s+/g, '');
    };
    ReturnInfoFlagsBase.prototype.addProperties = function (properties) {
        if (properties instanceof Object) {
            for (var name_1 in properties) {
                if (properties.hasOwnProperty(name_1)) {
                    var value = properties[name_1];
                    if (typeof value === "boolean") {
                        this.setFlag(name_1, value, !value);
                    }
                    else {
                        this.setValue(name_1, value);
                    }
                }
            }
        }
    };
    ReturnInfoFlagsBase.prototype.setProperty = function (key, value, defaultValue) {
        if (value !== defaultValue) {
            this.data[key] = value;
        }
    };
    return ReturnInfoFlagsBase;
}());
exports.ReturnInfoFlagsBase = ReturnInfoFlagsBase;
var ReturnInfo = /** @class */ (function () {
    // Accepts all InfoFlags
    function ReturnInfo(_a) {
        if (_a === void 0) { _a = {}; }
        var _b = _a.articleInfo, articleInfo = _b === void 0 ? new ArticleInfoFlags() : _b, _c = _a.eventInfo, eventInfo = _c === void 0 ? new EventInfoFlags() : _c, _d = _a.sourceInfo, sourceInfo = _d === void 0 ? new SourceInfoFlags() : _d, _e = _a.categoryInfo, categoryInfo = _e === void 0 ? new CategoryInfoFlags() : _e, _f = _a.conceptInfo, conceptInfo = _f === void 0 ? new ConceptInfoFlags() : _f, _g = _a.locationInfo, locationInfo = _g === void 0 ? new LocationInfoFlags() : _g, _h = _a.storyInfo, storyInfo = _h === void 0 ? new StoryInfoFlags() : _h, _j = _a.conceptClassInfo, conceptClassInfo = _j === void 0 ? new ConceptClassInfoFlags() : _j, _k = _a.conceptFolderInfo, conceptFolderInfo = _k === void 0 ? new ConceptFolderInfoFlags() : _k, unsupported = __rest(_a, ["articleInfo", "eventInfo", "sourceInfo", "categoryInfo", "conceptInfo", "locationInfo", "storyInfo", "conceptClassInfo", "conceptFolderInfo"]);
        if (Object.keys(unsupported).length > 0) {
            logger_1.Logger.warn("ReturnInfo: Unsupported parameters detected: ".concat(JSON.stringify(unsupported), ". Please check the documentation."));
        }
        this.articleInfo = articleInfo;
        this.eventInfo = eventInfo;
        this.sourceInfo = sourceInfo;
        this.categoryInfo = categoryInfo;
        this.conceptInfo = conceptInfo;
        this.locationInfo = locationInfo;
        this.storyInfo = storyInfo;
        this.conceptClassInfo = conceptClassInfo;
        this.conceptFolderInfo = conceptFolderInfo;
    }
    ReturnInfo.prototype.getParams = function (prefix) {
        return __assign(__assign(__assign(__assign(__assign(__assign(__assign(__assign(__assign({}, this.articleInfo.getProperties(prefix)), this.eventInfo.getProperties(prefix)), this.sourceInfo.getProperties(prefix)), this.categoryInfo.getProperties(prefix)), this.conceptInfo.getProperties(prefix)), this.locationInfo.getProperties(prefix)), this.storyInfo.getProperties(prefix)), this.conceptClassInfo.getProperties(prefix)), this.conceptFolderInfo.getProperties(prefix));
    };
    /**
     * load the configuration for the ReturnInfo from a filename
     * @param filename filename that contains the json configuration to use in the ReturnInfo
     */
    ReturnInfo.loadFromFile = function (filename) {
        if (!(fs && fs.existsSync(filename))) {
            throw new Error("File ".concat(filename, " does not exist"));
        }
        var conf = JSON.parse(fs.readFileSync(filename, "utf8"));
        return new ReturnInfo({
            articleInfo: new ArticleInfoFlags(conf.articleInfo || {}),
            eventInfo: new EventInfoFlags(conf.eventInfo || {}),
            sourceInfo: new SourceInfoFlags(conf.sourceInfo || {}),
            categoryInfo: new CategoryInfoFlags(conf.categoryInfo || {}),
            conceptInfo: new ConceptInfoFlags(conf.conceptInfo || {}),
            locationInfo: new LocationInfoFlags(conf.locationInfo || {}),
            storyInfo: new StoryInfoFlags(conf.storyInfo || {}),
            conceptClassInfo: new ConceptClassInfoFlags(conf.conceptClassInfo || {}),
            conceptFolderInfo: new ConceptFolderInfoFlags(conf.conceptFolderInfo || {}),
        });
    };
    return ReturnInfo;
}());
exports.ReturnInfo = ReturnInfo;
var ArticleInfoFlags = /** @class */ (function (_super) {
    __extends(ArticleInfoFlags, _super);
    function ArticleInfoFlags(params) {
        if (params === void 0) { params = {}; }
        var _this = _super.call(this) || this;
        _this.type = "Article";
        var _a = params.bodyLen, bodyLen = _a === void 0 ? -1 : _a, _b = params.basicInfo, basicInfo = _b === void 0 ? true : _b, _c = params.title, title = _c === void 0 ? true : _c, _d = params.body, body = _d === void 0 ? true : _d, _e = params.url, url = _e === void 0 ? true : _e, _f = params.eventUri, eventUri = _f === void 0 ? true : _f, _g = params.authors, authors = _g === void 0 ? true : _g, _h = params.concepts, concepts = _h === void 0 ? false : _h, _j = params.categories, categories = _j === void 0 ? false : _j, _k = params.links, links = _k === void 0 ? false : _k, _l = params.videos, videos = _l === void 0 ? false : _l, _m = params.image, image = _m === void 0 ? true : _m, _o = params.socialScore, socialScore = _o === void 0 ? false : _o, _p = params.sentiment, sentiment = _p === void 0 ? true : _p, _q = params.location, location = _q === void 0 ? false : _q, _r = params.extractedDates, extractedDates = _r === void 0 ? false : _r, _s = params.originalArticle, originalArticle = _s === void 0 ? false : _s, _t = params.storyUri, storyUri = _t === void 0 ? false : _t, properties = __rest(params, ["bodyLen", "basicInfo", "title", "body", "url", "eventUri", "authors", "concepts", "categories", "links", "videos", "image", "socialScore", "sentiment", "location", "extractedDates", "originalArticle", "storyUri"]);
        _this.setValue("bodyLen", bodyLen);
        _this.setFlag("basicInfo", basicInfo);
        _this.setFlag("title", title);
        _this.setFlag("body", body);
        _this.setFlag("url", url);
        _this.setFlag("eventUri", eventUri);
        _this.setFlag("authors", authors);
        _this.setFlag("concepts", concepts);
        _this.setFlag("categories", categories);
        _this.setFlag("links", links);
        _this.setFlag("videos", videos);
        _this.setFlag("image", image);
        _this.setFlag("socialScore", socialScore);
        _this.setFlag("sentiment", sentiment);
        _this.setFlag("location", location);
        _this.setFlag("extractedDates", extractedDates);
        _this.setFlag("originalArticle", originalArticle);
        _this.setFlag("storyUri", storyUri);
        _this.addProperties(properties);
        return _this;
    }
    return ArticleInfoFlags;
}(ReturnInfoFlagsBase));
exports.ArticleInfoFlags = ArticleInfoFlags;
var StoryInfoFlags = /** @class */ (function (_super) {
    __extends(StoryInfoFlags, _super);
    function StoryInfoFlags(params) {
        if (params === void 0) { params = {}; }
        var _this = _super.call(this) || this;
        _this.type = "Story";
        var _a = params.basicStats, basicStats = _a === void 0 ? true : _a, _b = params.location, location = _b === void 0 ? true : _b, _c = params.date, date = _c === void 0 ? false : _c, _d = params.title, title = _d === void 0 ? false : _d, _e = params.summary, summary = _e === void 0 ? false : _e, _f = params.concepts, concepts = _f === void 0 ? false : _f, _g = params.categories, categories = _g === void 0 ? false : _g, _h = params.medoidArticle, medoidArticle = _h === void 0 ? false : _h, _j = params.infoArticle, infoArticle = _j === void 0 ? false : _j, _k = params.commonDates, commonDates = _k === void 0 ? false : _k, _l = params.socialScore, socialScore = _l === void 0 ? false : _l, _m = params.imageCount, imageCount = _m === void 0 ? 0 : _m, properties = __rest(params, ["basicStats", "location", "date", "title", "summary", "concepts", "categories", "medoidArticle", "infoArticle", "commonDates", "socialScore", "imageCount"]);
        _this.setFlag("basicStats", basicStats);
        _this.setFlag("location", location);
        _this.setFlag("date", date);
        _this.setFlag("title", title);
        _this.setFlag("summary", summary);
        _this.setFlag("concepts", concepts);
        _this.setFlag("categories", categories);
        _this.setFlag("medoidArticle", medoidArticle);
        _this.setFlag("infoArticle", infoArticle);
        _this.setFlag("commonDates", commonDates);
        _this.setFlag("socialScore", socialScore);
        _this.setValue("imageCount", imageCount);
        _this.addProperties(properties);
        return _this;
    }
    return StoryInfoFlags;
}(ReturnInfoFlagsBase));
exports.StoryInfoFlags = StoryInfoFlags;
var EventInfoFlags = /** @class */ (function (_super) {
    __extends(EventInfoFlags, _super);
    function EventInfoFlags(params) {
        if (params === void 0) { params = {}; }
        var _this = _super.call(this) || this;
        _this.type = "Event";
        var _a = params.title, title = _a === void 0 ? true : _a, _b = params.summary, summary = _b === void 0 ? true : _b, _c = params.articleCounts, articleCounts = _c === void 0 ? true : _c, _d = params.concepts, concepts = _d === void 0 ? true : _d, _e = params.categories, categories = _e === void 0 ? true : _e, _f = params.date, date = _f === void 0 ? true : _f, _g = params.commonDates, commonDates = _g === void 0 ? false : _g, _h = params.infoArticle, infoArticle = _h === void 0 ? false : _h, _j = params.stories, stories = _j === void 0 ? false : _j, _k = params.socialScore, socialScore = _k === void 0 ? false : _k, _l = params.imageCount, imageCount = _l === void 0 ? 0 : _l, properties = __rest(params, ["title", "summary", "articleCounts", "concepts", "categories", "date", "commonDates", "infoArticle", "stories", "socialScore", "imageCount"]);
        _this.setFlag("title", title);
        _this.setFlag("summary", summary);
        _this.setFlag("articleCounts", articleCounts);
        _this.setFlag("concepts", concepts);
        _this.setFlag("categories", categories);
        _this.setFlag("date", date);
        _this.setFlag("commonDates", commonDates);
        _this.setFlag("infoArticle", infoArticle);
        _this.setFlag("stories", stories);
        _this.setFlag("socialScore", socialScore);
        _this.setValue("imageCount", imageCount);
        _this.addProperties(properties);
        return _this;
    }
    return EventInfoFlags;
}(ReturnInfoFlagsBase));
exports.EventInfoFlags = EventInfoFlags;
var SourceInfoFlags = /** @class */ (function (_super) {
    __extends(SourceInfoFlags, _super);
    function SourceInfoFlags(params) {
        if (params === void 0) { params = {}; }
        var _this = _super.call(this) || this;
        var _a = params.title, title = _a === void 0 ? true : _a, _b = params.description, description = _b === void 0 ? false : _b, _c = params.location, location = _c === void 0 ? false : _c, _d = params.ranking, ranking = _d === void 0 ? false : _d, _e = params.image, image = _e === void 0 ? false : _e, _f = params.socialMedia, socialMedia = _f === void 0 ? false : _f, properties = __rest(params, ["title", "description", "location", "ranking", "image", "socialMedia"]);
        _this.type = "Source";
        _this.setFlag("title", title);
        _this.setFlag("description", description);
        _this.setFlag("location", location);
        _this.setFlag("ranking", ranking);
        _this.setFlag("image", image);
        _this.setFlag("socialMedia", socialMedia);
        _this.addProperties(properties);
        return _this;
    }
    return SourceInfoFlags;
}(ReturnInfoFlagsBase));
exports.SourceInfoFlags = SourceInfoFlags;
var CategoryInfoFlags = /** @class */ (function (_super) {
    __extends(CategoryInfoFlags, _super);
    function CategoryInfoFlags(params) {
        if (params === void 0) { params = {}; }
        var _this = _super.call(this) || this;
        _this.type = "Category";
        var _a = params.trendingScore, trendingScore = _a === void 0 ? true : _a, properties = __rest(params, ["trendingScore"]);
        _this.setFlag("trendingScore", trendingScore);
        _this.addProperties(properties);
        return _this;
    }
    return CategoryInfoFlags;
}(ReturnInfoFlagsBase));
exports.CategoryInfoFlags = CategoryInfoFlags;
var ConceptInfoFlags = /** @class */ (function (_super) {
    __extends(ConceptInfoFlags, _super);
    function ConceptInfoFlags(params) {
        if (params === void 0) { params = {}; }
        var _this = _super.call(this) || this;
        _this.type = "Concept";
        var _a = params.type, type = _a === void 0 ? "concepts" : _a, _b = params.lang, lang = _b === void 0 ? "eng" : _b, _c = params.label, label = _c === void 0 ? true : _c, _d = params.synonyms, synonyms = _d === void 0 ? false : _d, _e = params.image, image = _e === void 0 ? false : _e, _f = params.description, description = _f === void 0 ? false : _f, _g = params.trendingScore, trendingScore = _g === void 0 ? false : _g, _h = params.maxConceptsPerType, maxConceptsPerType = _h === void 0 ? 20 : _h, properties = __rest(params, ["type", "lang", "label", "synonyms", "image", "description", "trendingScore", "maxConceptsPerType"]);
        _this.setValue("type", type);
        _this.setValue("lang", lang);
        _this.setFlag("label", label);
        _this.setFlag("synonyms", synonyms);
        _this.setFlag("image", image);
        _this.setFlag("description", description);
        _this.setFlag("trendingScore", trendingScore);
        _this.setValue("maxConceptsPerType", maxConceptsPerType, 20, true);
        _this.addProperties(properties);
        return _this;
    }
    return ConceptInfoFlags;
}(ReturnInfoFlagsBase));
exports.ConceptInfoFlags = ConceptInfoFlags;
var LocationInfoFlags = /** @class */ (function (_super) {
    __extends(LocationInfoFlags, _super);
    function LocationInfoFlags(params) {
        if (params === void 0) { params = {}; }
        var _this = _super.call(this) || this;
        _this.type = "Location";
        var _a = params.label, label = _a === void 0 ? true : _a, _b = params.wikiUri, wikiUri = _b === void 0 ? false : _b, _c = params.geoNamesId, geoNamesId = _c === void 0 ? false : _c, _d = params.population, population = _d === void 0 ? false : _d, _e = params.geoLocation, geoLocation = _e === void 0 ? false : _e, _f = params.countryArea, countryArea = _f === void 0 ? false : _f, _g = params.countryDetails, countryDetails = _g === void 0 ? false : _g, _h = params.countryContinent, countryContinent = _h === void 0 ? false : _h, _j = params.placeFeatureCode, placeFeatureCode = _j === void 0 ? false : _j, _k = params.placeCountry, placeCountry = _k === void 0 ? true : _k, properties = __rest(params, ["label", "wikiUri", "geoNamesId", "population", "geoLocation", "countryArea", "countryDetails", "countryContinent", "placeFeatureCode", "placeCountry"]);
        _this.setFlag("label", label);
        _this.setFlag("wikiUri", wikiUri);
        _this.setFlag("geoNamesId", geoNamesId);
        _this.setFlag("population", population);
        _this.setFlag("geoLocation", geoLocation);
        _this.setFlag("countryArea", countryArea);
        _this.setFlag("countryDetails", countryDetails);
        _this.setFlag("countryContinent", countryContinent);
        _this.setFlag("placeFeatureCode", placeFeatureCode);
        _this.setFlag("placeCountry", placeCountry);
        _this.addProperties(properties);
        return _this;
    }
    return LocationInfoFlags;
}(ReturnInfoFlagsBase));
exports.LocationInfoFlags = LocationInfoFlags;
var ConceptClassInfoFlags = /** @class */ (function (_super) {
    __extends(ConceptClassInfoFlags, _super);
    function ConceptClassInfoFlags(params) {
        if (params === void 0) { params = {}; }
        var _this = _super.call(this) || this;
        _this.type = "ConceptClass";
        var _a = params.parentLabels, parentLabels = _a === void 0 ? true : _a, _b = params.concepts, concepts = _b === void 0 ? false : _b, properties = __rest(params, ["parentLabels", "concepts"]);
        _this.setFlag("parentLabels", parentLabels);
        _this.setFlag("concepts", concepts);
        _this.addProperties(properties);
        return _this;
    }
    return ConceptClassInfoFlags;
}(ReturnInfoFlagsBase));
exports.ConceptClassInfoFlags = ConceptClassInfoFlags;
var ConceptFolderInfoFlags = /** @class */ (function (_super) {
    __extends(ConceptFolderInfoFlags, _super);
    function ConceptFolderInfoFlags(params) {
        if (params === void 0) { params = {}; }
        var _this = _super.call(this) || this;
        _this.type = "ConceptFolder";
        var _a = params.definition, definition = _a === void 0 ? true : _a, _b = params.owner, owner = _b === void 0 ? false : _b, properties = __rest(params, ["definition", "owner"]);
        _this.setFlag("owner", definition);
        _this.setFlag("owner", owner);
        _this.addProperties(properties);
        return _this;
    }
    return ConceptFolderInfoFlags;
}(ReturnInfoFlagsBase));
exports.ConceptFolderInfoFlags = ConceptFolderInfoFlags;
